### ISSUE TITLE
Usage of hard set paths displaying host information leading to future nspawn container failures.

### ISSUE THREAD

**Issue by oiaohm** [systemd/systemd#16416]
**systemd version the issue has been seen with**
https://github.com/systemd/systemd/pull/15891/files
Above is merged leading to following in master.
https://github.com/systemd/systemd/blob/976efe1b80864199b190573d938467dab3f33661/src/nspawn/nspawn-mount.c#L566

**Expected behaviour you didn't see**
The ability to define "/run/host/etc/os-release" and "/run/host/usr/lib/os-release" to point to somewhere other than  "/run/host/etc/os-release"  and "/usr/lib/os-release"

**Unexpected behaviour you saw**

No ability to change this.    Fixing this will need some serous thoughts for security reasons.   Like allowing /etc/os-release.d directory to contain different os-release personalities files that can be defines by nspawn mount process to be "/run/host/etc/os-release"  and "/run/host/usr/lib/os-release" instead of the hosts os-release.      This could go as far as no os-release.d directory equals do not expose  "/run/host/etc/os-release"  and "/run/host/usr/lib/os-release" at all. 

There is no real reason why /etc/os-release"  and "/usr/lib/os-release" could not be symlinked as "/etc/os-release.d/host-etc" and "/etc/os-release.d/host-lib" and these used as the defaults if person administrating system does not have these files or has deleted them then tell any container about os-release is disabled.

Web browsers include means to fake user-agent for very good reasons you hit sites that incorrectly check.      This will happen in production soon enough in production if this is not fixed their will not be a work around without modify the container.   Some cases the part checking could be a binary blob application that administrator of system cannot change.

**Steps to reproduce the problem**

Create a nspawn container or portable service that checks 
"/run/host/usr/lib/os-release" or "/run/host/etc/os-release" for expected value that does not match the host that automatically fails because of it.   

Result you are after is that container can check for something other than the host values see them and succeed.

**Options**
I understand there are reasons why people want to be able to include information about host for debugging reasons and the like but there also has to be ability to lie to container what host is to allow for those creating containers being stupidly restrictive..

I understand the arguement that since uname gives you information a little more is not a problem but that is not exactly true.
mount --bind ./cpuinfo.raspi3 /proc/cpuinfo like solutions you already find when people are using docker container to cross build stuff to lie about CPU to contained runtime so stuff builds correctly.

This is why I said general hard set paths problem there are items in containers that are being passed through now that will lead to container failures.   Like having the ability to have a Raspberry pi 3 personality mask for particular containers would be good for building applications on a x86 for a Raspberry pi 3 even better if these personality settings don't have to be done on a per unit file base.    A personality mask of of some form could also altering the way nspawn setups mounts before starting containers as well and fix the os-release problem as well.

This is a problem that needs fixing before we have people using this feature wrong so when they do it will be easy to fix.   We can 100 percent sure that someone making container running software is going to check this new feature using a method that the program crashes or fails when os-release contains information that did not test with.   Fake this information need to exist by some clean means.

**Comment by poettering** [comment]
not sure i follow, but with --bind-ro= you can neatly overmount the stuff with anything you like?

**Comment by oiaohm** [comment]
> not sure i follow, but with --bind-ro= you can neatly overmount the stuff with anything you like?

Yes you can over bind.   But this will lead to errors  with the current patch as the wrong value are exposed.

https://github.com/systemd/systemd/pull/15891/files#diff-dfea18b372a7f0b032fe089756381e18R124

If I just override "/run/host/etc/os-release" now the $container_host_id value does not match because it still processed from /etc/os-release.

https://github.com/systemd/systemd/pull/15891/files#diff-db6b16c9ade7e951e3bb793bec110098R3305   This here is using direct host /etc/os-release.

Really you need to be able to set where in host "/etc/os-release 
 and "/usr/lib/os-release" in fact point to populate  the container_host_id and other stuff.

I would also suspect by the time --bind-ro= stuff is processed its also too late as the $container_host_id and its other values are already populated.

Basically there are two things that need to be in absolute sync.   The environment variable of the container_host stuff and the displayed info in "/run/host/usr/lib/os-release" and "/run/host/etc/os-release"  so a simple bind mount over the files does not fix it unless the population of the container_host stuff is done after all the bind mounts is complete from the "/run/host/usr/lib/os-release" or "/run/host/etc/os-release" not directly from the host /etc/os-release yes this is not how that patch works.   The patch accepted is not design to allow me as administrator to override os-release info completely because the patch design is leaking host information into environmental vars that bind mount is not going to fix.

Like it or not that patch was not designed with the idea that administrator would need to provide false os-release information in container with the files and the environmental vars that the container can see all lining up perfectly.

Need to be very careful leaking information about host into containers always need to be done so administrator can fake it or turn it off completely if required.

**Comment by poettering** [comment]
/cc @bluca 

I'd be happy to make the host path of the os-release file to us configurable via an env var or so, i.e. so that you can call `SYSTEMD_NSPAWN_OVERRIDE_OS_RELEASE=/etc/fake-os-release systemd-nspawn`...

Such a patch should be trivial, and I'd be happy to review/merge it. WOuldn't that suffice for you?

**Comment by bluca** [comment]
It's already simple to override both, without patches though?
Just use ```systemd-nspawn --set-env=container_host_id=foobar``` for the env var and ```--bind-ro=/dev/null:/run/host/os-release``` for the file (or adjust as desired). I made the feature intentionally apply the default values first, so that users can override them easily if they wish to do so.

But note:

> The patch accepted is not design to allow me as administrator to override os-release info completely because the patch design is leaking host information into environmental vars that bind mount is not going to fix.

As explained many times, there's nothing that wasn't already available that gets exposed via os-release - eg, from /proc:

```
cat /proc/sys/kernel/version
#1 SMP Debian 5.6.14-2~bpo10+1 (2020-06-09)
```

The kernel version allows a much finer grained identification of the host system, as it pin-points a small window during which a kernel package was released and uploaded. From os-release you only get the distro major version.

**Comment by bluca** [comment]
Of course if someone wants to add additional shortcuts, it's perfectly fine by me
